File: multipaz-compose/src/commonMain/kotlin/org/multipaz/compose/document/DocumentModel.kt
package org.multipaz.compose.document

import androidx.compose.ui.graphics.ImageBitmap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.io.bytestring.ByteString
import org.multipaz.cbor.Cbor
import org.multipaz.cbor.DataItem
import org.multipaz.cbor.buildCborMap
import org.multipaz.cbor.putCborMap
import org.multipaz.compose.branding.Branding
import org.multipaz.compose.decodeImage
import org.multipaz.credential.Credential
import org.multipaz.credential.SecureAreaBoundCredential
import org.multipaz.crypto.Algorithm
import org.multipaz.crypto.Crypto
import org.multipaz.document.Document
import org.multipaz.document.DocumentAdded
import org.multipaz.document.DocumentDeleted
import org.multipaz.document.DocumentEvent
import org.multipaz.document.DocumentStore
import org.multipaz.document.DocumentUpdated
import org.multipaz.documenttype.DocumentTypeRepository
import org.multipaz.storage.Storage
import org.multipaz.storage.StorageTable
import org.multipaz.storage.StorageTableSpec
import org.multipaz.storage.ephemeral.EphemeralStorage
import org.multipaz.util.Logger
import org.multipaz.util.LruCache

private const val TAG = "DocumentModel"

private data class DocumentModelStorageData(
    var sortingOrder: Map<String, Int> = emptyMap()
) {
    fun toDataItem(): DataItem {
        return buildCborMap {
            putCborMap("documentOrder") {
                for((key, value) in sortingOrder) {
                    put(key, value)
                }
            }
        }
    }

    suspend fun save(
        table: StorageTable,
        partitionId: String
    ) {
        table.update(
            key = KEY_NAME,
            data = ByteString(Cbor.encode(toDataItem())),
            partitionId = partitionId
        )
    }

    companion object {
        const val KEY_NAME = "DocumentModelStorageData"

        suspend fun load(
            table: StorageTable,
            partitionId: String
        ): DocumentModelStorageData {
            table.get(KEY_NAME, partitionId)?.let {
                return fromDataItem(Cbor.decode(it.toByteArray()))
            }
            val data = DocumentModelStorageData()
            table.insert(KEY_NAME, ByteString(Cbor.encode(data.toDataItem())), partitionId)
            return data
        }

        fun fromDataItem(dataItem: DataItem): DocumentModelStorageData {
            var sortingOrder = emptyMap<String, Int>()
            try {
                if (dataItem.hasKey("documentOrder")) {
                    sortingOrder = dataItem["documentOrder"].asMap.map { (key, value) ->
                        key.asTstr to value.asNumber.toInt()
                    }.toMap()
                }
            } catch (e: Throwable) {
                Logger.e(TAG, "Error decoding sortingOrder", e)
            }
            return DocumentModelStorageData(sortingOrder)
        }
    }
}

/**
 * Model that loads documents from a [DocumentStore] and keeps them updated.
 *
 * This model exposes a [StateFlow] of all documents as [DocumentInfo]
 * and listens to live updates from the store. If a [Document] has no card art the model
 * creates a default card art using [Branding.renderFallbackCardArt]. The model also
 * maintains a persistent order of documents and applications can call e.g.
 * [setDocumentPosition] to change the order.
 *
 * @param documentStore the [DocumentStore] which manages [Document] and [Credential] instances.
 * @param documentTypeRepository a [DocumentTypeRepository] with information about document types or `null`.
 * @param storage the [Storage] used for storing document order.
 * @param storagePartition the partition of [storage] to use.
 */
class DocumentModel(
    val documentStore: DocumentStore,
    val documentTypeRepository: DocumentTypeRepository?,
    val storage: Storage = EphemeralStorage(),
    val storagePartition: String = "default",
) {
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default)
    private val _documentInfos = MutableStateFlow<List<DocumentInfo>>(emptyList())
    private lateinit var table: StorageTable
    private lateinit var storageData: DocumentModelStorageData

    /**
     * A list of [DocumentInfo] for the documents in [documentStore].
     */
    val documentInfos: StateFlow<List<DocumentInfo>> = _documentInfos.asStateFlow()

    init {
        scope.launch {
            table = storage.getTable(documentModelTableSpec)
            storageData = DocumentModelStorageData.load(table, storagePartition)

            val docIds = documentStore.listDocumentIds()
            docIds.forEach { documentId ->
                updateDocumentInfo(documentId, DocumentAdded(documentId))
            }

            documentStore.eventFlow
                .onEach { event ->
                    Logger.i(
                        TAG,
                        "DocumentStore event ${event::class.simpleName} ${event.documentId}"
                    )
                    updateDocumentInfo(event = event)

                }
                .launchIn(scope)
        }
    }

    private suspend fun updateDocumentInfo(
        documentId: String? = null,
        event: DocumentEvent? = null,
    ) {
        val id = event?.documentId ?: documentId ?: return
        when (event) {
            is DocumentAdded -> {
                documentStore.lookupDocument(id)?.let { document ->
                    _documentInfos.update { current ->
                        current
                            .toMutableList()
                            .apply {
                                add(document.toDocumentInfo())
                            }.sorted()
                    }
                }
            }
            is DocumentDeleted -> {
                _documentInfos.update { current ->
                    current
                        .toMutableList()
                        .apply {
                            find { documentInfo -> documentInfo.document.identifier == id }?.let {
                                remove(it)
                            }
                        }.sorted()
                }
            }
            is DocumentUpdated -> {
                documentStore.lookupDocument(id)?.let { document ->
                    _documentInfos.update { current ->
                        current
                            .toMutableList()
                            .apply {
                                val existingDocumentInfo =
                                    find { documentInfo -> documentInfo.document.identifier == id }
                                if (existingDocumentInfo == null) {
                                    Logger.w(TAG, "Didn't find DocumentInfo for document with id $id")
                                } else {
                                    val newDocumentInfo = document.toDocumentInfo()
                                    if (newDocumentInfo != existingDocumentInfo) {
                                        remove(existingDocumentInfo)
                                        add(newDocumentInfo)
                                    } else {
                                        Logger.w(TAG, "DocumentInfo for document with id $id didn't change")
                                    }
                                }
                            }.sorted()
                    }
                }
            }
            null -> {}
        }
    }

    private fun List<DocumentInfo>.sorted(): List<DocumentInfo> {
        return this.sortedWith { a, b ->
            val sa = storageData.sortingOrder[a.document.identifier]
            val sb = storageData.sortingOrder[b.document.identifier]
            if (sa != null && sb != null) {
                if (sa != sb) {
                    return@sortedWith sa.compareTo(sb)
                }
            }
            return@sortedWith a.document.created.compareTo(b.document.created)
        }
    }

    /**
     * Sets the position of a document.
     *
     * @param documentInfo the [DocumentInfo] to set the position for.
     * @param position the new position, zero-based.
     * @throws IllegalArgumentException if [documentInfo] doesn't exist in the model or if [position]
     *   exceed the number of documents in the store.
     */
    suspend fun setDocumentPosition(
        documentInfo: DocumentInfo,
        position: Int
    ) {
        val documentInfos = _documentInfos.value.toMutableList()
        if (!documentInfos.remove(documentInfo)) {
            throw IllegalArgumentException("Passed in documentInfo is not in list")
        }
        documentInfos.add(
            index = position,
            element = documentInfo,
        )
        val sortingOrder = mutableMapOf<String, Int>()
        documentInfos.forEachIndexed { index, documentInfo ->
            sortingOrder.put(documentInfo.document.identifier, index)
        }
        storageData.sortingOrder = sortingOrder
        storageData.save(table, storagePartition)
        _documentInfos.value = documentInfos
    }

    // Use a simple LRU cache to avoid decoding the same cardArt over and over again
    private val cardArtCache = LruCache<ByteString, ImageBitmap>(5)

    private suspend fun Document.toDocumentInfo(): DocumentInfo {
        cardArt?.let {
            val image = it.toByteArray()
            val imageSha256 = ByteString(Crypto.digest(Algorithm.SHA256, image))
            var cardArt = cardArtCache.get(imageSha256)
            if (cardArt == null) {
                cardArt = decodeImage(image)
                cardArtCache.put(imageSha256, cardArt)
            }
            return DocumentInfo(
                document = this,
                cardArt = cardArt,
                credentialInfos = buildCredentialInfos(documentTypeRepository)
            )
        }
        return DocumentInfo(
            document = this,
            cardArt = Branding.Current.value.renderFallbackCardArt(this),
            credentialInfos = buildCredentialInfos(documentTypeRepository)
        )
    }

    companion object {
        private val documentModelTableSpec = StorageTableSpec(
            name = "DocumentModel",
            supportPartitions = true,
            supportExpiration = false,
        )

        private suspend fun Document.buildCredentialInfos(
            documentTypeRepository: DocumentTypeRepository?
        ): List<CredentialInfo> {
            return getCredentials().map { credential ->
                val keyInfo = if (credential is SecureAreaBoundCredential) {
                    credential.secureArea.getKeyInfo(credential.alias)
                } else {
                    null
                }
                val keyInvalidated = if (credential is SecureAreaBoundCredential) {
                    credential.secureArea.getKeyInvalidated(credential.alias)
                } else {
                    false
                }
                val claims = if (credential.isCertified) {
                    credential.getClaims(documentTypeRepository)
                } else {
                    emptyList()
                }
                CredentialInfo(
                    credential = credential,
                    claims = claims,
                    keyInfo = keyInfo,
                    keyInvalidated = keyInvalidated
                )
            }
        }
    }
}

File: multipaz-swift/Sources/MultipazSwift/DocumentModel.swift
@preconcurrency import Multipaz
import UIKit
import Combine

/// A structure with information about a ``Credential``.
public struct CredentialInfo: Hashable {
    /// A reference to the ``Credential`` this information is about.
    public let credential: Credential
    
    /// The claims in the credential.
    public let claims: [Claim]
    
    /// Information about the key-binding key if the credential is a ``SecureAreaBoundCredential``.
    public let keyInfo: KeyInfo?
    
    /// True if the credential is a ``SecureAreaBoundCredential`` and the key has been invalidated.
    public let keyInvalidated: Bool
}

/// A structure with information about a ``Document``.
public struct DocumentInfo: Hashable {
    /// A reference to the ``Document`` this information is about.
    public let document: Document
    
    /// Card art for the document.
    public let cardArt: UIImage
    
    /// The credentials for the document.
    public let credentialInfos: [CredentialInfo]
    
    public static func == (lhs: DocumentInfo, rhs: DocumentInfo) -> Bool {
        return lhs.document.identifier == rhs.document.identifier
    }
}

/**
 * Errors that can be thrown by ``DocumentModel``.
 */
public enum DocumentModelError: Error {
    case noSuchDocument
    case positionOutOfRange
}

/**
 * Model that loads documents from a ``DocumentStore`` and keeps them updated.
 *
 * The model exposes the documents as ``DocumentInfo`` and listens to live updates from the store
 * and maintains a persistent order which can be changed using ``setDocumentPosition(documentInfo:position:)``.
 *
 * If a ``Document`` has no cardArt the model creates a default stock cardArt.
 */
@MainActor
@Observable
public class DocumentModel {
    
    let documentTypeRepository: DocumentTypeRepository?
    let storage: Storage
    let storagePartition: String

    /**
     * Initialization for ``DocumentModel``.
     *
     * - Parameters:
     *   - documentTypeRepository: a ``DocumentTypeRepository`` with information about document types or nil.
     *   - storage: the [Storage] used for storing document order.
     *   - storagePartition: the partition of [storage] to use.
     */
    public init(
        documentTypeRepository: DocumentTypeRepository?,
        storage: Storage = EphemeralStorage(clock: KotlinClockCompanion.shared.getSystem()),
        storagePartition: String = "default"
    ) {
        self.documentTypeRepository = documentTypeRepository
        self.storage = storage
        self.storagePartition = storagePartition
    }
    
    private var _documentInfos: [DocumentInfo] = []

    public var documentInfos: [DocumentInfo] {
        _documentInfos.sorted { (a: DocumentInfo, b: DocumentInfo) -> Bool in
            let sa = storageData.sortingOrder[a.document.identifier]
            let sb = storageData.sortingOrder[b.document.identifier]
            if sa != nil && sb != nil {
                if sa != sb {
                    return sa! < sb!
                }
            }
            return Document.Comparator.shared.compare(a: a.document, b: b.document) < 0
        }
    }

    private var documentStore: DocumentStore!
    private var table: StorageTable!
    private var storageData: DocumentModelStorageData!
    
    private let documentModelTableSpec = StorageTableSpec(
        name: "DocumentModel",
        supportPartitions: true,
        supportExpiration: false,
        schemaVersion: 0
    )

    public func setDocumentStore(documentStore: DocumentStore) async {
        self.documentStore = documentStore
        self.table = try! await storage.getTable(spec: documentModelTableSpec)
        self.storageData = try! await DocumentModelStorageData.load(
            table: self.table,
            partitionId: self.storagePartition
        )

        for document in try! await documentStore.listDocuments(sort: true) {
            await _documentInfos.append(getDocumentInfo(document))
        }
        Task {
            for await event in documentStore.eventFlow {
                if event is DocumentAdded {
                    let document = try! await documentStore.lookupDocument(identifier: event.documentId)
                    if document != nil {
                        await self._documentInfos.append(getDocumentInfo(document!))
                    }
                } else if event is DocumentUpdated {
                    let index = self._documentInfos.firstIndex { documentInfo in
                        documentInfo.document.identifier == event.documentId
                    }
                    if (index != nil) {
                        self._documentInfos[index!] = await getDocumentInfo(self._documentInfos[index!].document)
                    }
                } else if event is DocumentDeleted {
                    self._documentInfos.removeAll { documentInfo in
                        documentInfo.document.identifier == event.documentId
                    }
                }
            }
        }
    }
    
    /**
     * Sets the position of a document.
     *
     * - Parameters:
     *  - documentInfo: the ``DocumentInfo`` to set position for.
     *  - position: the position to set.
     * - Throws: ``DocumentError.noSuchDocument`` if the given ``DocumentInfo`` doesn't exist.
     * - Throws: ``DocumentError.positionOutOfRange`` if the given position is out of range.
     */
    public func setDocumentPosition(
        documentInfo: DocumentInfo,
        position: Int
    ) async throws {
        var documentInfos = self.documentInfos
        let index = documentInfos.firstIndex(of: documentInfo)
        if index == nil {
            throw DocumentModelError.noSuchDocument
        }
        documentInfos.remove(at: index!)
        if position < 0 || position > documentInfos.count {
            throw DocumentModelError.positionOutOfRange
        }
        documentInfos.insert(documentInfo, at: position)
        var sortingOrder: [String:Int] = [:]
        documentInfos.enumerated().forEach { index, di in
            sortingOrder[di.document.identifier] = index
        }
        storageData = DocumentModelStorageData(sortingOrder: sortingOrder)
        try! await storageData.save(table: table, partitionId: storagePartition)
    }

    private func getDocumentInfo(_ document: Document) async -> DocumentInfo {
        var credentialInfos: [CredentialInfo] = []
        for credential in try! await document.getCredentials() {
            await credentialInfos.append(getCredentialInfo(credential))
        }
        return DocumentInfo(
            document: document,
            cardArt: document.renderCardArt(),
            credentialInfos: credentialInfos
        )
    }

    private func getCredentialInfo(_ credential: Credential) async -> CredentialInfo {

        var keyInfo: KeyInfo? = nil
        var keyInvalidated = false
        if let secureAreaBoundCredential = credential as? SecureAreaBoundCredential {
            keyInfo = try! await secureAreaBoundCredential.secureArea.getKeyInfo(alias: secureAreaBoundCredential.alias)
            keyInvalidated = try! await secureAreaBoundCredential.isInvalidated().boolValue
        }
        let claims: [Claim] = if credential.isCertified {
            try! await credential.getClaims(documentTypeRepository: documentTypeRepository)
        } else {
            []
        }
        return CredentialInfo(
            credential: credential,
            claims: claims,
            keyInfo: keyInfo,
            keyInvalidated: keyInvalidated
        )
    }
}



fileprivate struct DocumentModelStorageData {
    private static let keyName = "DocumentModelStorageData"
    
    var sortingOrder: [String: Int] = [:]
    
    func toDataItem() -> DataItem {
        let builder = CborMap.companion.builder()
        let innerBuilder = builder.putMap(key: Tstr(value: "documentOrder"))
        for (key, value) in sortingOrder {
            innerBuilder.put(
                key: Tstr(value: key),
                value: value >= 0 ? Uint(value: UInt64(value)) : Nint(value: UInt64(value))
            )
        }
        return builder.end()!.build()
    }
    
    func save(table: StorageTable, partitionId: String) async throws {
        let data = try Cbor.shared.encode(item: toDataItem())
        
        try await table.update(
            key: Self.keyName,
            data: ByteString(data: data, startIndex: 0, endIndex: data.size),
            partitionId: partitionId,
            expiration: nil
        )
    }
    
    static func load(table: StorageTable, partitionId: String) async throws -> DocumentModelStorageData {
        // Try to get existing data
        if let data = try await table.get(key: keyName, partitionId: partitionId) {
            let dataItem = try Cbor.shared.decode(encodedCbor: data.toByteArray(startIndex: 0, endIndex: data.size))
            return fromDataItem(dataItem)
        }
        
        // If not found, create new, insert, and return
        let newData = DocumentModelStorageData()
        let encodedData = try Cbor.shared.encode(item: newData.toDataItem())
        
        try await table.insert(
            key: keyName,
            data: ByteString(data: encodedData, startIndex: 0, endIndex: encodedData.size),
            partitionId: partitionId,
            expiration: KotlinInstant.companion.DISTANT_FUTURE
        )
        
        return newData
    }
    
    static func fromDataItem(_ dataItem: DataItem) -> DocumentModelStorageData {
        var sortingOrder: [String: Int] = [:]
        if dataItem.hasKey(key: "documentOrder") {
            for (key, value) in dataItem.get(key: "documentOrder").asMap {
                sortingOrder[key.asTstr] = Int(value.asNumber)
            }
        }
        return DocumentModelStorageData(sortingOrder: sortingOrder)
    }
}
